File: Android/TUIKit/TUICallKit/tuicallkit/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/utils/UserInfoUtils.java
package com.tencent.qcloud.tuikit.tuicallkit.utils;

import android.text.TextUtils;

import com.tencent.imsdk.v2.V2TIMFriendInfo;
import com.tencent.imsdk.v2.V2TIMFriendInfoResult;
import com.tencent.imsdk.v2.V2TIMManager;
import com.tencent.imsdk.v2.V2TIMValueCallback;
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.TUILog;
import com.tencent.qcloud.tuikit.tuicallkit.base.CallingUserModel;

import java.util.ArrayList;
import java.util.List;

public class UserInfoUtils {
    private static final String TAG                  = "UserInfoUtils";
    private static final String ERROR_MSG_EMPTY_LIST = "userList is empty";

    public UserInfoUtils() {
    }

    public void getUserInfo(String userId, final UserCallback callback) {
        if (TextUtils.isEmpty(userId)) {
            TUILog.e(TAG, "getUserInfo, userId is empty.");
            if (callback != null) {
                callback.onFailed(-1, ERROR_MSG_EMPTY_LIST);
            }
            return;
        }

        List<String> list = new ArrayList<>();
        list.add(userId);
        getUserInfoByIdList(list, callback);
    }

    public void getUserInfo(List<CallingUserModel> userIdList, final UserCallback callback) {
        if (null == userIdList || userIdList.isEmpty()) {
            TUILog.e(TAG, "getUserInfo, userIdList is empty.");
            if (callback != null) {
                callback.onFailed(-1, ERROR_MSG_EMPTY_LIST);
            }
            return;
        }

        List<String> list = new ArrayList<>();
        for (CallingUserModel model : userIdList) {
            if (null != model && !TextUtils.isEmpty(model.userId)) {
                list.add(model.userId);
            }
        }
        getUserInfoByIdList(list, callback);
    }

    //Return the remark first.If remark is empty, return the nickname. If nickname is also empty, return the userId.
    public void getUserInfoByIdList(List<String> userList, UserCallback callback) {
        if (null == userList || userList.isEmpty()) {
            TUILog.e(TAG, "getUserInfo, userIdList is empty");
            if (callback != null) {
                callback.onFailed(-1, ERROR_MSG_EMPTY_LIST);
            }
            return;
        }
        V2TIMManager.getFriendshipManager().getFriendsInfo(userList,
                new V2TIMValueCallback<List<V2TIMFriendInfoResult>>() {
                    @Override
                    public void onError(int errorCode, String errorMsg) {
                        TUILog.e(TAG, "getUserInfo failed, errorCode: " + errorCode + ", errorMsg: " + errorMsg);
                        if (callback != null) {
                            callback.onFailed(errorCode, errorMsg);
                        }
                    }

                    @Override
                    public void onSuccess(List<V2TIMFriendInfoResult> v2TIMFriendInfoResults) {
                        if (v2TIMFriendInfoResults == null || v2TIMFriendInfoResults.isEmpty()) {
                            TUILog.e(TAG, "getUserInfo result is empty");
                            if (null != callback) {
                                callback.onFailed(-1, ERROR_MSG_EMPTY_LIST);
                            }
                            return;
                        }

                        List<CallingUserModel> list = new ArrayList<>();
                        for (V2TIMFriendInfoResult result : v2TIMFriendInfoResults) {
                            if (result == null) {
                                continue;
                            }
                            CallingUserModel model = new CallingUserModel();

                            V2TIMFriendInfo friendInfo = result.getFriendInfo();
                            model.userId = friendInfo.getUserID();
                            model.userName = friendInfo.getFriendRemark();

                            if (TextUtils.isEmpty(model.userName)) {
                                String nickName = friendInfo.getUserProfile().getNickName();
                                model.userName = TextUtils.isEmpty(nickName) ? friendInfo.getUserID() : nickName;
                            }

                            model.userAvatar = friendInfo.getUserProfile().getFaceUrl();
                            if (!isUrl(model.userAvatar)) {
                                model.userAvatar = "";
                            }
                            list.add(model);
                        }
                        TUILog.i(TAG, "getUserInfo, list: " + list);
                        if (callback != null) {
                            callback.onSuccess(list);
                        }
                    }
                });
    }

    private boolean isUrl(String url) {
        return !TextUtils.isEmpty(url) && (url.startsWith("http://") || url.startsWith("https://"));
    }

    public interface UserCallback {
        void onSuccess(List<CallingUserModel> list);

        void onFailed(int errorCode, String errorMsg);
    }
}

File: Android/TUIKit/TUICallKit/tuicallkit-kt/src/main/java/com/tencent/qcloud/tuikit/tuicallkit/utils/UserInfoUtils.kt
package com.tencent.qcloud.tuikit.tuicallkit.utils

import android.text.TextUtils
import com.tencent.imsdk.v2.V2TIMFriendInfoResult
import com.tencent.imsdk.v2.V2TIMManager
import com.tencent.imsdk.v2.V2TIMValueCallback
import com.tencent.qcloud.tuikit.TUICommonDefine.ValueCallback
import com.tencent.qcloud.tuikit.tuicallengine.impl.base.TUILog
import com.tencent.qcloud.tuikit.tuicallkit.data.User

object UserInfoUtils {

    private const val TAG = "UserInfoUtils"

    fun updateUserInfo(user: User) {
        if (TextUtils.isEmpty(user.id)) {
            TUILog.e(TAG, "getUsersInfo -> user.userId isEmpty")
            return
        }
        if (!TextUtils.isEmpty(user.nickname.get()) && !TextUtils.isEmpty(user.avatar.get())) {
            TUILog.i(
                TAG,
                "getUsersInfo -> user.userName = ${user.nickname}, avatar = ${user.avatar}"
            )
            return
        }
        val userList: MutableList<String> = ArrayList()
        userList.add(user.id!!)

        getFriendsInfo(userList, object : ValueCallback<List<UserInfo?>?> {
            override fun onSuccess(data: List<UserInfo?>?) {
                if (data.isNullOrEmpty() || data[0] == null) {
                    TUILog.e(TAG, "getUserInfo result is empty")
                    return
                }
                var userInfo = data[0]
                TUILog.i(
                    TAG, "getUsersInfo -> userId:${user.id} onSuccess: " +
                            "nickname:${userInfo?.nickname}" +
                            ", avatar:${userInfo?.avatar}" +
                            ", friendRemark:${userInfo?.remark}"
                )
                if (!TextUtils.isEmpty(userInfo?.remark)) {
                    user.nickname.set(userInfo?.remark)
                } else if (!TextUtils.isEmpty(userInfo?.nickname)) {
                    user.nickname.set(userInfo?.nickname)
                } else {
                    user.nickname.set(userInfo?.id)
                }
                user.avatar.set(userInfo?.avatar)
            }

            override fun onError(errCode: Int, errMsg: String?) {
                TUILog.e(
                    TAG,
                    "getUsersInfo -> userId:${user.id} onError errorCode = $errCode , errorMsg = $errMsg"
                )
            }

        })
    }

    fun getUserListInfo(userList: List<String>, callback: ValueCallback<List<User>?>) {
        getFriendsInfo(userList, object : ValueCallback<List<UserInfo?>?> {
            override fun onSuccess(data: List<UserInfo?>?) {
                if (data.isNullOrEmpty()) {
                    TUILog.e(TAG, "getUserInfo result is empty")
                    return
                }

                var userList: MutableList<User> = ArrayList()
                for (i in data.indices) {
                    var userInfo = data[i]

                    var user = User()
                    user.id = userInfo?.id
                    if (!TextUtils.isEmpty(userInfo?.remark)) {
                        user.nickname.set(userInfo?.remark)
                    } else if (!TextUtils.isEmpty(userInfo?.nickname)) {
                        user.nickname.set(userInfo?.nickname)
                    } else {
                        user.nickname.set(userInfo?.id)
                    }
                    user.avatar.set(userInfo?.avatar)
                    userList.add(user)
                }
                callback?.onSuccess(userList)
            }

            override fun onError(errCode: Int, errMsg: String?) {
                TUILog.e(
                    TAG,
                    "getUsersInfo -> userId:${userList} onError errorCode = $errCode , errorMsg = $errMsg"
                )
                callback.onError(errCode, errMsg)
            }

        })
    }

    private fun getFriendsInfo(
        userList: List<String?>?,
        callback: ValueCallback<List<UserInfo?>?>?
    ) {
        V2TIMManager.getFriendshipManager()
            .getFriendsInfo(userList, object : V2TIMValueCallback<List<V2TIMFriendInfoResult?>?> {
                override fun onSuccess(list: List<V2TIMFriendInfoResult?>?) {
                    if (list.isNullOrEmpty()) {
                        TUILog.e(TAG, "getUserInfo result is empty")
                        return
                    }

                    var userList: MutableList<UserInfo> = ArrayList()
                    for (i in list.indices) {
                        var friendInfo = list[i]?.friendInfo

                        var userInfo = UserInfo()
                        userInfo.id = friendInfo?.userID ?: ""
                        userInfo.remark = friendInfo?.friendRemark ?: ""
                        userInfo.nickname = friendInfo?.userProfile?.nickName ?: ""
                        userInfo.avatar = friendInfo?.userProfile?.faceUrl ?: ""
                        userList.add(userInfo)
                    }
                    callback?.onSuccess(userList)
                }

                override fun onError(code: Int, desc: String?) {
                    callback?.onError(code, desc)
                }
            })
    }

    class UserInfo {
        var id: String = ""
        var avatar: String = ""
        var nickname: String = ""
        var remark: String = ""
    }
}